core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -233,13 +235,13 @@ public Balancer(ESLogger logger, RoutingAllocation allocation, WeightFunction we
             this.threshold = threshold;
             this.routingNodes = allocation.routingNodes();
             this.metaData = allocation.metaData();
-            float totalShardWeight = 0.0f;
+            int totalShardWeight = 0;
             for (IndexMetaData indexMetaData : allocation.metaData()) {
-                float shardWeightMultiplier = INDEX_SHARD_WEIGHT_MULTIPLIER_SETTING.get(indexMetaData.getSettings());
-                shardWeightMultipliers.put(indexMetaData.getIndex().getName(), shardWeightMultiplier);
-                totalShardWeight += shardWeightMultiplier * indexMetaData.getTotalNumberOfShards();
+                int shardWeight = INDEX_BALANCE_SHARD_WEIGHT.get(indexMetaData.getSettings());
+                shardWeights.put(indexMetaData.getIndex().getName(), shardWeight);
+                totalShardWeight += shardWeight * indexMetaData.getTotalNumberOfShards();
```

Comment:
should we check here that the totalShardWeight is not negative. I was just thinking if somebody uses the number of docs per shard and we overflow? I really wonder if we should put some upperbound into the setting to ensure folk don't go crazy? They should use some log scale rather than actual numbers? maybe we use `1<<16` as the upper limit for now? and move totalShardWeight to a long and use doubles elsewhere? I really just wanna protect us form going negative :)